include/kodlab_mjbots_sdk/loop_cache.h

```diff
@@ -0,0 +1,103 @@
+/**
+ * @file loop_cache.h
+ * @author J. Diego Caporale (jdcap@seas.upenn.edu)
+ * @brief Control loop validated cache and id for automatic invalidation synced 
+ * with loop
+ * @date 2023-03-30
+ * 
+ * @copyright Copyright (c) 2023 The Trustees of the University of Pennsylvania. 
+ * All Rights Reserved
+ * BSD 3-Clause License
+ */
+
+#pragma once
+#include <type_traits>
+#include <memory>
+
+#include "kodlab_mjbots_sdk/common_header.h"
+
+namespace kodlab{
+
+/**
+ * @brief Simple loop id counter class that allows the user to tell when the 
+ * control loop has looped. Useful for forcing recalculations and/or timing.
+ * 
+ */
+class LoopId
+{
+ public:
+  /**
+   * @brief Returns the current loop id
+   * 
+   * @return uint32_t 
+   */
+  static uint32_t get(){
+    return loop_id_;
+  }
+
+  /**
+   * @brief Override the current loop id
+   * 
+   * @param id_des 
+   */
+  static void set(uint32_t id_des){
+    loop_id_ = id_des;
+  }
+
+  /**
+   * @brief increment and return the loop id, note 
+   * @note there can only be ONE loop using this type of cache at once to avoid 
+   * prematurely invalidating caches, MAKE SURE THIS IS WHAT YOU WANT
+   * 
+   * @return uint32_t 
+   */
+  static uint32_t increment(){
+      return (++loop_id_);
+  }
+
+ protected:
+  /**
+   * @brief Loop id shared between all instances acts a "global" loop id
+   * 
+   */
+  static uint32_t loop_id_; 
+};
```

Comment:
This is a mutable singleton with some possible issues.

First, philosophically, why use a singleton here?  It goes against the [ISO guidelines](http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#i3-avoid-singletons), and the resulting behavior could instead be maintained in a local scope.  E.g., a loop ID is maintained in the control loop and all `ValdatedLoopCache`s are constructed with a reference/pointer to that ID (and can use the same invalidation logic). 

That said, since this is a pretty high-level, integral thing, then a singleton could be fine.  But:
- this singleton could suffer from the [static initialization order fiasco](https://stackoverflow.com/a/35958423) if `LoopId` were to be referenced in another static variable initialization;
- `loop_id_` can be arbitrarily modified by children of `LoopId`;
- `LoopId` can be multiply constructed (which could be confusing).

You could alleviate some of these issues by making `LoopId` a [Meyer's](https://stackoverflow.com/a/1008289) [Singleton](https://laristra.github.io/flecsi/src/developer-guide/patterns/meyers_singleton.html).  I think something like this would work.

```cpp
class LoopId {
 public:
  static LoopId& get() {
    static LoopId loop_id;
    return loop_id;
  }

  static void set(uint32_t id) {
    auto loop_id = LoopId::get();
    loop_id = id;
  }

  static void increment() {
    auto loop_id = LoopId::get();
    loop_id++;
  }

  LoopId(const LoopId&) = delete;
  LoopId& operator=(const LoopId&) = delete;

private:
  LoopId() {}
  ~LoopId() {}

};
```